src/rust/engine/logging/src/executor.rs

```diff
@@ -0,0 +1,130 @@
+  /// The future will be driven to completion, but the result can't be accessed directly.
+  ///
+  /// This may be useful e.g. if you want to kick off a potentially long-running task, which will
+  /// notify dependees of its completion over an mpsc channel.
+  ///
+  pub fn spawn_and_ignore<F: Future<Item = (), Error = ()> + Send + 'static>(&self, future: F) {
+    self
+      .runtime
+      .executor()
+      .spawn(Self::future_with_correct_logging_context(future))
+  }
+
+  ///
+  /// Run a Future on a tokio Runtime as a new Task, and return a Future handle to it.
+  ///
+  /// The future will only be driven to completion if something drives the returned Future. If the
+  /// returned Future is dropped, the computation may be cancelled.
+  ///
+  /// This may be useful for tokio tasks which use the tokio blocking feature (unrelated to the
+  /// Executor::block_on method). When tokio blocking tasks run, they prevent progress on any
+  /// futures running in the same task. e.g. if you run f1.select(f2) and f1 and f2 are
+  /// tokio blocking futures, f1 and f2 will not run in parallel, defeating the point of select.
+  ///
+  /// On the other hand, if you run:
+  /// spawn_oneshot(f1).select(spawn_oneshot(f2))
+  /// those futures will run in parallel.
+  ///
+  /// Using spawn_oneshot allows for selecting the granularity when using tokio blocking.
+  ///
+  /// See https://docs.rs/tokio-threadpool/0.1.15/tokio_threadpool/fn.blocking.html for details of
+  /// tokio blocking.
+  ///
+  pub fn spawn_oneshot<
+    Item: Send + 'static,
+    Error: Send + 'static,
+    F: Future<Item = Item, Error = Error> + Send + 'static,
+  >(
+    &self,
+    future: F,
+  ) -> impl Future<Item = Item, Error = Error> {
+    futures::sync::oneshot::spawn(
+      Self::future_with_correct_logging_context(future),
+      &self.runtime.executor(),
+    )
+  }
+
+  ///
+  /// Run a Future and return its resolved Result.
+  ///
+  /// This should never be called from in a Future context, and should only ever be called in
+  /// something that resembles a main method.
+  ///
+  /// This method makes a new Runtime every time it runs, to ensure that the caller doesn't
+  /// accidentally deadlock by using this when a Future attempts to itself call
+  /// Executor::spawn_and_ignore or Executor::spawn_oneshot. Because it should be used only in very
+  /// limited situations, this overhead is viewed to be acceptable.
+  ///
+  pub fn block_on<
+    Item: Send + 'static,
+    Error: Send + 'static,
+    F: Future<Item = Item, Error = Error> + Send + 'static,
+  >(
+    &self,
+    future: F,
+  ) -> Result<Item, Error> {
+    // Make sure to copy our (thread-local) logging destination into the task.
+    // When a daemon thread kicks off a future, it should log like a daemon thread (and similarly
+    // for a user-facing thread).
+    Runtime::new()
+      .unwrap()
+      .block_on(Self::future_with_correct_logging_context(future))
+  }
+
+  pub fn spawn_on_io_pool<
```

Comment:
Is it worth discussing the path toward tokio-fs in a docstring here? Would the idea be to add a `spawn_io` with ~the same shape, that spawned a `blocking` call to a tokio-fs API? If so, would it be less churn to leave the dep in, and expose that method now with one (?) usage?

Reply:
Added a docstring with my imagined path. It's possible we'd end up swapping out the implementation of `spawn_on_io_pool`, but I imagine we'd probably end up wanting to call `spawn_oneshot` and do the blocking annotating either at the call-site, or in a helper method.

Definitely a problem for the future.